homeassistant/components/wiffi/__init__.py

```diff
@@ -0,0 +1,220 @@
+
+
+class WiffiIntegrationApi:
+    """API object for wiffi handling. Stored in hass.data."""
+
+    def __init__(self, hass):
+        """Initialize the instance."""
+        self._hass = hass
+        self._server = None
+        self._known_devices = {}
+        self._periodic_callback = None
+
+    def async_setup(self, config_entry):
+        """Set up api instance."""
+        self._server = WiffiTcpServer(config_entry.data[CONF_PORT], self)
+        self._periodic_callback = async_track_time_interval(
+            self._hass, self._periodic_tick, timedelta(seconds=10)
+        )
+
+    def shutdown(self):
+        """Shutdown wiffi api.
+
+        Remove listener for periodic callbacks.
+        """
+        remove_listener = self._periodic_callback
+        if remove_listener is not None:
+            remove_listener()
+
+    async def __call__(self, device, metrics):
+        """Process callback from TCP server if new data arrives from a device."""
+        if device.mac_address not in self._known_devices:
+            # add empty set for new device
+            self._known_devices[device.mac_address] = set()
+
+        for metric in metrics:
+            if metric.id not in self._known_devices[device.mac_address]:
+                self._known_devices[device.mac_address].add(metric.id)
+                async_dispatcher_send(self._hass, CREATE_ENTITY_SIGNAL, device, metric)
+            else:
+                async_dispatcher_send(
+                    self._hass,
+                    f"{UPDATE_ENTITY_SIGNAL}-{generate_unique_id(device, metric)}",
+                    device,
+                    metric,
+                )
+
+    @property
+    def server(self):
+        """Return TCP server instance for start + close."""
+        return self._server
+
+    @callback
+    def _periodic_tick(self, now=None):
+        """Check if any entity has timed out because it has not been updated."""
+        async_dispatcher_send(self._hass, CHECK_ENTITIES_SIGNAL)
+
+
+class WiffiEntity(Entity):
+    """Common functionality for all wiffi entities."""
+
+    def __init__(self, device, metric):
+        """Initialize the base elements of a wiffi entity."""
+        self._id = generate_unique_id(device, metric)
+        self._device_info = {
+            "connections": {
+                (device_registry.CONNECTION_NETWORK_MAC, device.mac_address)
+            },
+            "identifiers": {(DOMAIN, device.mac_address)},
+            "manufacturer": "stall.biz",
+            "name": f"{device.moduletype} {device.mac_address}",
+            "model": device.moduletype,
+            "sw_version": device.sw_version,
+        }
+        self._name = metric.description
+        self._expiration_date = None
+        self._value = None
+
+    async def async_added_to_hass(self):
+        """Entity has been added to hass."""
+        async_dispatcher_connect(
+            self.hass, f"{UPDATE_ENTITY_SIGNAL}-{self._id}", self._update_value_callback
+        )
```

Comment:
Use `self.async_on_remove` to unsubscribe the dispatcher.
```suggestion
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass, f"{UPDATE_ENTITY_SIGNAL}-{self._id}", self._update_value_callback
            )
        )
```